src/strokeMatches.ts

```diff
@@ -20,34 +20,44 @@ const START_AND_END_DIST_THRESHOLD = 250; // bigger = more lenient
 const FRECHET_THRESHOLD = 0.4; // bigger = more lenient
 const MIN_LEN_THRESHOLD = 0.35; // smaller = more lenient
 
+export type StrokeMatchType = 'miss' | 'match' | 'backwards-match';
```

Comment:
Cool, I'm happy to change `allowBackwardsStrokes` to boolean and to return an object from `strokeMatches`.

I like the idea of `failureReasons`, but are the `*Match` variables independent enough that they can be used to convey that level of specificity about failure? This may be my unfamiliarity with the code, but when I was playing around with it, I wasn't able to determine whether a stroke was valid aside from being backward just by looking at the existing vars. For instance, with the `identifies backwards stroke when allowed` test, the initial check has `shapeMatch: false` and `startAndEndMatch: true` which both seem incorrect (or at least unintuitive) to me. This is why I went the the approach of recursing with the reversed points array.

<img width="352" alt="Screen Shot 2021-09-24 at 10 28 48 PM" src="https://user-images.githubusercontent.com/8584728/134754933-82990887-2dbd-4656-97b1-febdca4e6792.png">

Reply:
Yeah you're doing it right I think. The shape match is using Frechet Distance, which is dependent on direction already. This is a good point, it's hard to isolate a single reason why a stroke failed, because it could fail for multiple reasons, and the reasons sort of overlap (Frechet distance is distance + shape + direction, for example).

Maybe something like this in that case?
```
type StrokeMatchResult {
  match: boolean; // whether or not the match is successful
  failureMeta?: {
    strokeIsBackwards: boolean // if the stroke would have been accepted if it wasn't backwards
    // in the future we can add more stuff here if we want
  }
}
```